packages/components/src/toolbar/toolbar-button/index.tsx

```diff
@@ -1,33 +1,37 @@
-// @ts-nocheck
-
 /**
  * External dependencies
  */
 import classnames from 'classnames';
+import type { ForwardedRef } from 'react';
+
 /**
  * WordPress dependencies
  */
 import { useContext, forwardRef } from '@wordpress/element';
+
 /**
  * Internal dependencies
  */
 import Button from '../../button';
 import ToolbarItem from '../toolbar-item';
 import ToolbarContext from '../toolbar-context';
 import ToolbarButtonContainer from './toolbar-button-container';
+import type { ToolbarButtonProps } from './types';
+import type { WordPressComponentProps } from '../../ui/context';
+import type React from 'react';
 
-function ToolbarButton(
+function UnforwardedToolbarButton(
 	{
-		containerClassName,
+		children,
 		className,
+		containerClassName,
 		extraProps,
-		children,
-		title,
 		isActive,
 		isDisabled,
+		title,
 		...props
-	},
-	ref
+	}: WordPressComponentProps< ToolbarButtonProps, typeof Button, false >,
```

Comment:
Good question. Considering this from several angles:

- Can Emotion handle polymorphism like this? — Yes. The Emotion `as` prop allows `React.ElementType`, so it _can_ potentially handle `as={ Button }` if this was a polymorphic WordPressComponent.
- Is it a good idea to allow something like `as={ Button }`? — In general, I don't like it because it's hard to predict what will happen. But it may make sense in some abstract components.
- Should we prefer `ToolbarButtonProps & ButtonProps` syntax in cases like this? — No strong opinion, given that there are no serious downsides to the `WordPressComponentProps` syntax. I like how we can just do `typeof Button` without importing the `ButtonProps` type though.

Reply:
Good observations. Let's keep as-is, keep an eye for any unintended consequences, and otherwise consider it as another tool in our belt :)